src/net_processing.cpp

```diff
@@ -253,12 +257,19 @@ namespace {
             return &(*a) < &(*b);
         }
     };
-    std::map<COutPoint, std::set<std::map<uint256, COrphanTx>::iterator, IteratorComparator>> mapOrphanTransactionsByPrev GUARDED_BY(g_cs_orphans);
 
-    std::vector<std::map<uint256, COrphanTx>::iterator> g_orphan_list GUARDED_BY(g_cs_orphans); //! For random eviction
+    /** Index from COutPoint into the mapOrphanTransactions. Used to remove
+     *  orphan transactions from the mapOrphanTransactions */
+    std::map<COutPoint, std::set<std::map<uint256, COrphanTx>::iterator, IteratorComparator>> mapOrphanTransactionsByPrev GUARDED_BY(g_cs_orphans);
+    /** Orphan transactions listed in random order for random eviction */
```

Comment:
1/3 nit on this comment, 2/3 for my understanding- 
these orphans aren't listed exactly random right? `AddOrphanTx` adds the entries to in the order they are seen, but `EraseOrphanTx` introduces some randomness bc it deletes an entry by swapping out the last element (which I'm still trying to understand the point of). 

seems the point of this data structure is to enable `LimitOrphanTxSize` to easily choose a random orphan to erase by providing a vector interface? 

if my understanding is correct, I think it'd be clearer to say "List of orphan transactions to enable random eviction"

Reply:
> these orphans aren't listed exactly random right?

> if my understanding is correct, I think it'd be clearer to say "List of orphan transactions to enable random eviction"

You're correct. I've now fixed the comment. Thank you!

> it deletes an entry by swapping out the last element (which I'm still trying to understand the point of)

This is because `g_orphan_list` is a std::vector, which is a contiguous-memory container. It can't have gaps in the underlying array. The normal way to erase an element from the middle of the vector is to delete it and then shift all the following elements forward by one position. That's expensive, and would invalidate all of the `list_pos` values that point back to a position in the vector. It's more efficient to copy the last element to the position of the element you want to delete, fix up the `list_pos` of just that `COrphanTx`, and then pop of the (now duplicate) last element.

Reply:
does this sound right: 

it seems to me` that `list_pos` is tracked _for this reason_. so the alternative would be if we removed `list_pos` & used in the built in vector erase function. the tradeoff would be less efficiency. 

the current design choice is to store this additional field on every `COrphanTx` to reduce an O(n) operation on the `g_orphan_list` vector, where n <= 100 | `-maxorphantx`. but I don't have much idea of how often we erase orphans in practice. so small increase in memory for small decrease in processing time?

Reply:
Yes, this is the only reason we store `list_pos`. Without it, we'd have to do a (relatively) expensive erase from the middle of the vector.